client/genesis-builder/src/lib.rs

```diff
@@ -0,0 +1,47 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Substrate genesis config builder
+//!
+//! This Runtime API allows to construct `GenesisConfig`, in particular:
+//! - serialize the runtime default `GenesisConfig` struct into json format,
+//! - put the GenesisConfig struct into the storage. Internally this operation calls
+//!   `GenesisBuild::build` function
+//! for all runtime pallets, which is typically by pallet's author.
+//! - deserialize the GenesisConfig from given json blob and put GenesisConfig into the state
+//!   storage. Allows to build
+//! customized configuration.
+//!
+//! Providing externalities with empty storage and putting GenesisConfig into storage allows to
+//! catch and build the raw storage of `GenesisConfig` which is the foundation for genesis block.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+sp_api::decl_runtime_apis! {
```

Comment:
@bkchr ,
maybe dumb question, but I am not sure what is good decision here:

When drafting GenesisBuilder Runtime API (https://github.com/paritytech/substrate/pull/14131)  I've put there some functions which work on `json`.
IMO it would be good to return [`serde_json::error`](https://github.com/serde-rs/json/blob/187f7dadc6ab3423128dde5e5598a83617fb18f2/src/error.rs#L175-L189) (`Result` [here](https://github.com/serde-rs/json/blob/187f7dadc6ab3423128dde5e5598a83617fb18f2/src/error.rs#L15-L23)).
But in order to use this type while crossing _runtime border_, it must implement `TypeInfo`.

I see following options:
- do some wrapper for `serde_json::error`, that copies the fields (and assert sizes) and implements the `TypeInfo + From/Into(serde_json::error)`.
- the other approach could be String, but from what I see `format` is rather not used in runtime.

What do you think?

Reply:
Invalid `GenesisConfig` JSON should be quite rare. I think just using an error `enum` that has an `JsonError` variant is enough.  
I dont think that we need to return the full serde error here.